### PR TITLE
Use Lwt's sequence-associated storage instead of our own Thread_local

### PR DESCRIPTION
**Pull request by ELLIOTTCABLE**
This stores the implicit 'global' surrounding-scope off the callstack in Lwt's "[sequence-associated storage][]" instead of our own Thread_local, as this makes more sense in the context of green threads.

I tried to match the code-style, let me know if you have any tweaks.

   [sequence-associated storage]: <https://github.com/ocsigen/lwt/blob/5.4.1/src/core/lwt.ml#L739-L775>

### CODE REVIEW

**Review by c-cube — APPROVED** [review]
That seems great to me, and follows the conventions of the blocking code. @mattjbray your opinion?

**Comment by ELLIOTTCABLE** [comment]
This actually has non-obvious and subtle performance consequences — we've been talking about this internally at @Ahrefs for a bit. To quote @mfp,

> tldr
> * Lwt TLS works by associating a heterogeneous map to promises
> * If you naïvely spawn async threads with `ignore (make_some_promise ())` or equivalent (which we do in multiple places, and is wrong for multiple reasons  btw.), you're saving the whole map. Lwt has no way to know which keys will be used in subsequent lookups.
> * This can lead to leaks if you have long-running async threads that are launched in the middle of a promise with a heavy `Lwt.with_value` map.

Basically, it makes "using Lwt wrong in a mostly-harmless way, suddenly harmful." And 'it' in this case is a supposedly-transparent debugging and tracing library. Ew. I think we'll continue with this approach internally, but we have a more total control than an open-source library can have …

… unfortunately, _not_ making this change is also problematic. (Actual)threads are complicated in a Lwt context … I'm not sure how to proceed.

If we merge this, it should be with some _significant_ documentation, educating users up-front about how to be more principled and careful with their promises … idk. Thoughts?
